### PR TITLE
llama-context: add ability to get logits

### PR DESCRIPTION
**Pull request by am17an**
From #14644

### CODE REVIEW

File: src/llama-context.cpp [hunk 1]
```diff
@@ -791,10 +792,22 @@ int llama_context::encode(const llama_batch & batch_inp) {
...
+        GGML_ASSERT(logits != nullptr);
+
+        if (n_outputs) {
+            ggml_backend_tensor_get_async(backend_res, t_logits, logits, 0, n_outputs*n_vocab*sizeof(float));
+        }
+    }
+
     // extract embeddings
-    if (t_embd) {
+    if (cparams.embeddings && t_embd) {
```
Comment (ggerganov) [inline src/llama-context.cpp]:
```suggestion
    if (embd && t_embd) {
```

File: src/llama-context.cpp [hunk 2]
```diff
@@ -791,10 +792,22 @@ int llama_context::encode(const llama_batch & batch_inp) {
...
+   if (t_logits && n_outputs > 0) {
+        ggml_backend_t backend_res = ggml_backend_sched_get_tensor_backend(sched.get(), t_logits);
+        GGML_ASSERT(backend_res != nullptr);
+        GGML_ASSERT(logits != nullptr);
+
+        if (n_outputs) {
+            ggml_backend_tensor_get_async(backend_res, t_logits, logits, 0, n_outputs*n_vocab*sizeof(float));
+        }
+    }
+
```
Comment (ggerganov) [inline src/llama-context.cpp]:
```suggestion
    // extract logits
   if (logits && t_logits) {
        ggml_backend_t backend_res = ggml_backend_sched_get_tensor_backend(sched.get(), t_logits);
        GGML_ASSERT(backend_res != nullptr);
        GGML_ASSERT(logits != nullptr);

        ggml_backend_tensor_get_async(backend_res, t_logits, logits, 0, n_tokens*n_vocab*sizeof(float));
    }

```

**Review by ggerganov — APPROVED** [review]
